runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -54,10 +56,25 @@
   // TODO: (BEAM-723) Create a shared ExecutorService for maintenance tasks in the DirectRunner.
   private static final ExecutorService COUNTER_COMMITTER = Executors.newCachedThreadPool();
 
-  private interface MetricAggregation<UpdateT, ResultT> {
-    UpdateT zero();
-    UpdateT combine(Iterable<UpdateT> updates);
-    ResultT extract(UpdateT data);
+  private static class MetricAggregation<UpdateT extends MetricData<ResultT>, ResultT> {
+    UpdateT initialValue;
+
+    public MetricAggregation(UpdateT zero) {
+      this.initialValue = zero;
+    }
+
+    @SuppressWarnings("unchecked")
+    UpdateT combine(Iterable<UpdateT> updates) {
+      UpdateT value = initialValue;
+      for (UpdateT update : updates) {
```

Comment:
This puts a boxed, virtual dispatch and arithmetic in what was a primitive addition. I'm somewhat worried about this from a performance perspective. I believe this has been previously benchmarked, and is why (for instance) the combineFn code for SumIntCombineFn is written the way it does (and why it uses int[] instead of Integer).